crates/runtime/src/dataconnector/graphql.rs

```diff
@@ -0,0 +1,285 @@
+            let buf = v.to_string();
+            let batch = ReaderBuilder::new(Arc::clone(&schema))
+                .with_batch_size(1024)
+                .build(Cursor::new(buf.as_bytes()))
+                .context(UnableToReadBatchSnafu)?
+                .collect::<Result<Vec<_>, _>>()
+                .context(UnableToCollectBatchSnafu)?;
+            res.push(batch);
+        }
+
+        Ok((res, schema))
+    }
+}
+
+#[allow(clippy::needless_pass_by_value)]
+fn to_execution_error(e: Error) -> DataFusionError {
+    DataFusionError::Execution(format!("{e}"))
+}
+
+impl GraphQL {
+    fn get_client(&self, dataset: &Dataset) -> super::DataConnectorResult<GraphQLClient> {
+        let mut client_builder = reqwest::Client::builder();
+        let token = get_secret_or_param(&self.params, &self.secret, "auth_token_key", "auth_token");
+        let user = get_secret_or_param(&self.params, &self.secret, "auth_user_key", "auth_user");
+        let pass = get_secret_or_param(&self.params, &self.secret, "auth_pass_key", "auth_pass");
+
+        let query = self
+            .params
+            .get("query")
+            .ok_or("Query not found in params".into())
+            .context(super::InvalidConfigurationSnafu {
+                dataconnector: "GraphQL",
+                message: "Query not found",
+            })?
+            .to_owned();
+        let endpoint = Url::parse(&dataset.path()).map_err(Into::into).context(
+            super::InvalidConfigurationSnafu {
+                dataconnector: "GraphQL",
+                message: "Invalid URL",
+            },
+        )?;
+        let json_path = self
+            .params
+            .get("json_path")
+            .map_or("data".to_string(), ToOwned::to_owned);
+
+        let mut headers = HeaderMap::new();
+        headers.append(USER_AGENT, HeaderValue::from_static("spice"));
+        headers.append(CONTENT_TYPE, HeaderValue::from_static("application/json"));
+
+        let mut auth = None;
+        if let Some(token) = token {
+            auth = Some(Auth::Bearer(token));
+        }
+        if let Some(user) = user {
+            auth = Some(Auth::Basic(user, pass));
+        }
+
+        client_builder = client_builder.default_headers(headers);
+
+        Ok(GraphQLClient {
+            client: client_builder.build().map_err(|e| {
+                super::DataConnectorError::InvalidConfiguration {
+                    dataconnector: "GraphQL".to_string(),
+                    message: "Failed to set token".to_string(),
+                    source: e.into(),
+                }
+            })?,
+            query,
+            endpoint,
+            json_path,
+            auth,
+        })
+    }
+}
+
+struct GraphQLTableProvider {
+    client: GraphQLClient,
+    schema: SchemaRef,
+}
+
+impl GraphQLTableProvider {
+    pub async fn new(client: GraphQLClient) -> super::DataConnectorResult<Self> {
+        let (_, schema) = client.execute().await.map_err(Into::into).context(
```

Comment:
Also here schema won't necessary require to load entire json response to infer to. As for federated query (no accelerator), this can be inferred by a default low value.

Reply:
Should be addressed with pagination and pushdown prs.